File: EternalAI.py/uniswap_ai/uniswap_ai/swap.py
import logging
import decimal

import requests
from web3 import Web3
from web3.middleware import geth_poa_middleware
from uniswap.constants import ETH_ADDRESS
from uniswap_ai.const import RPC_URL, ETH_CHAIN_ID
from dataclasses import dataclass
from uniswap import Uniswap

"""
You are an intelligent assistant specialized in cryptocurrency trading. When a user requests a transaction, please respond in a specific JSON format.

For example, if the user says: "Swap 1 ETH to USDT on Uniswap with Ethereum Chain(or Base chain, BNB chain...)" you should respond as follows:

{
  "token_in": "name of Input token",
  "token_in_amount": 1.0,
  "token_out": "Name of Ouput token",
}

Make sure that the token addresses are accurate and comply with ERC20 standards.
"""

"""
You are an intelligent assistant specialized in cryptocurrency trading. When a user requests a transaction, please respond in a specific JSON format.\n\nFor example, if the user says: \"Swap 1 ETH to USDT on Uniswap with Ethereum Chain (or Base chain, BNB chain...)\" you should respond as follows:\n\n{\n  \"token_in\": \"name of Input token\",\n  \"token_in_amount\": 1.0,\n  \"token_out\": \"Name of Output token\"\n}\n\nMake sure that the token addresses are accurate and comply with ERC20 standards.
"""


@dataclass
class SwapReq:
    token_in: str = ""
    token_in_amount: decimal = 0
    token_out: str = ""

    def convert_in_out(self):
        if self.token_in.lower() == "eth":
            self.token_in = ETH_ADDRESS
        else:
            token_address = self.convert_token_address(self.token_in.lower())
            if token_address is None:
                self.token_in = ETH_ADDRESS
            else:
                self.token_in = token_address

        token_address = self.convert_token_address(self.token_out.lower())
        if token_address is None:
            self.token_out = ETH_ADDRESS
        else:
            self.token_out = token_address

    def get_token_address_from_info(self, token_info: dict, symbol: str):
        result = None
        if token_info.get("symbol") != "" and token_info.get("symbol") is not None and token_info.get(
                "symbol") == symbol:
            if token_info.get("platforms") is not None:
                if token_info.get("platforms").get("ethereum") is not None and token_info.get("platforms").get(
                        "ethereum") != "":
                    result = token_info.get("platforms").get("ethereum")
        return result

    def convert_token_address(self, symbol: str):
        result = None
        token_info_response = requests.get(f"https://api.coingecko.com/api/v3/coins/{symbol}")
        if token_info_response.status_code == 200:
            result = self.get_token_address_from_info(token_info_response.json())
        if result is None:
            token_info_list_response = requests.get(
                "https://api.coingecko.com/api/v3/coins/list?include_platform=true&status=active")
            if token_info_list_response.status_code == 200:
                for token_info in token_info_list_response.json():
                    result = self.get_token_address_from_info(token_info, symbol)
                    if result is not None:
                        break

        return result


@dataclass
class UniSwapAI:
    web3: Web3 = None

    def create_web3(self, rpc: str = ""):
        if self.web3 is None:
            if rpc != "":
                self.web3 = Web3(Web3.HTTPProvider(rpc))
            else:
                # Default:
                self.web3 = Web3(Web3.HTTPProvider(RPC_URL[ETH_CHAIN_ID]))
            self.web3.middleware_onion.inject(geth_poa_middleware, layer=0)

    def swap_v3(self, privateKey: str, req: SwapReq, rpc: str = ""):
        if privateKey is None or len(privateKey) == 0:
            raise Exception('Private key missing')
        self.create_web3(rpc)
        if self.web3.is_connected():
            account = self.web3.eth.account.from_key(privateKey)
            account_address = Web3.to_checksum_address(account.address)

            """Checks price impact for a pool with liquidity."""
            uni_swap = Uniswap(address=account_address, private_key=privateKey, web3=self.web3, version=3)

            token_a = Web3.to_checksum_address(req.token_in)
            amount_in = self.web3.to_wei(req.token_in_amount, 'ether')

            token_b = Web3.to_checksum_address(req.token_out)

            # Compare the results with the output of:
            logging.info(f"https://app.uniswap.org/#/swap?use=v3&inputCurrency={token_a}&outputCurrency={token_b}")
            impact = uni_swap.estimate_price_impact(token_a, token_b, amount_in, fee=10000)
            logging.info(f"Impact for buying VXV on v3 with {amount_in / 10 ** 18} ETH:  {impact}")

            if token_a != ETH_ADDRESS:
                approved = uni_swap._is_approved(token_b)
                logging.info(f"approved: {approved}")

                if not approved:
                    logging.info("call approved")
                    uni_swap.approve(token_b, max_approval=None)

                    approved = uni_swap._is_approved(token_b)
                    logging.info(f"check approved again: {approved}")

            logging.info(f"call swap")
            tx = uni_swap.make_trade(token_a, token_b, amount_in, uni_swap.address, slippage=impact)
            logging.info(f"sending tx: {tx.to_checksum_address()}")
            return tx

    # def swap(self, privateKey: str, req: SwapReq):
    #     if self.web3 is None:
    #         self.web3 = Web3(Web3.HTTPProvider(rpc_url))
    #     if self.web3.is_connected():
    #         if privateKey is None or len(privateKey) == 0:
    #             privateKey = os.getenv("PRIVATE_KEY")
    #
    #         account = self.web3.eth.account.from_key(privateKey)
    #         account_address = account.address
    #
    #         amount_in = self.web3.to_wei(req.token_a_amount, 'ether')
    #         amount_out_min = self.web3.to_wei(req.token_b_out_min, 'ether')
    #         deadline = self.web3.eth.get_block('latest')['timestamp'] + 120
    #
    #         if self.uniswap_router_address == "":
    #             self.uniswap_router_address = "0xe592427a0aece92de3edee1f18e0157c05861564"
    #
    #         uniswap_router = self.web3.eth.contract(address=self.uniswap_router_address, abi=uniswap_router_abi)
    #         txn = uniswap_router.functions.swapExactTokensForTokens(
    #             amount_in,
    #             amount_out_min,
    #             [token_a, token_b],
    #             account.address,
    #             deadline
    #         ).buildTransaction({
    #             'from': account_address,
    #             'gas': 200000,
    #             # 'gasPrice': web3.toWei('50', 'gwei'),
    #             'nonce': self.web3.eth.get_transaction_count(account_address),
    #         })
    #
    #         signed_txn = self.web3.eth.account.sign_transaction(txn, private_key)
    #         txn_hash = self.web3.eth.send_raw_transaction(signed_txn.rawTransaction)
    #
    #         logging.info(f'Transaction hash: {self.web3.to_hex(txn_hash)}')
    #     else:
    #         raise Exception('Uniswap: web3 is not connected')

File: EternalAI.js/packages/uniswap/src/swap.ts
import {ETH_CHAIN_ID, getRPC, ZeroAddress} from "./const";
import {ethers} from "ethers";
import {Token} from "@uniswap/sdk-core";
import {changeWallet, createWallet, TransactionState} from "./libs/providers";
import {createTrade, executeTrade} from "./libs/trading";
import {CurrentConfig, Environment} from "./libs/config";
import {getCurrencyBalance, getCurrencyDecimal} from "./libs/wallet";

export class SwapReq {
    token_in: string = ""
    token_in_address: string = ""
    token_in_amount: number = 0
    token_out: string = ""
    token_out_address: string = ""

    constructor(token_in: string, token_in_address: string, token_in_amount: number, token_out: string, token_out_address: string) {
        this.token_in = token_in;
        this.token_in_address = token_in_address;
        this.token_in_amount = token_in_amount;
        this.token_out = token_out;
        this.token_out_address = token_out_address;
    }

    static fromJSON(json: string): SwapReq {
        const parsed = JSON.parse(json);
        return Object.assign(new SwapReq('', '', 0, '', ''), parsed);
    }

    convert_in_out = async () => {
        if (this.token_in.toLowerCase() == "eth") {
            this.token_in_address = ZeroAddress;
        } else {
            const token_address = await this.convert_token_address(this.token_in.toLowerCase())
            if (!token_address) this.token_in_address = ZeroAddress
            else this.token_in_address = token_address
        }

        if (this.token_out.toLowerCase() == "eth") {
            this.token_out_address = ZeroAddress;
        } else {
            const token_address = await this.convert_token_address(this.token_out.toLowerCase())
            if (!token_address) this.token_out_address = ZeroAddress
            else this.token_out_address = token_address
        }
    }

    get_token_address_from_info = (token_info: any, symbol: string) => {
        let result = null;
        if (token_info["symbol"] && token_info["symbol"] == symbol) {
            if (token_info["platforms"]) {
                if (token_info["platforms"]["ethereum"]) {
                    result = token_info["platforms"]["ethereum"]
                }
            }
        }
        return result;
    }

    convert_token_address = async (symbol: string) => {
        let result = null;
        const token_info_response = await fetch("https://api.coingecko.com/api/v3/coins/" + symbol)
        if (token_info_response.ok) {
            const data = await token_info_response.json();
            result = this.get_token_address_from_info(data, symbol)
        }
        if (!result) {
            const token_info_list_response = await fetch("https://api.coingecko.com/api/v3/coins/list?include_platform=true&status=active")
            if (token_info_list_response.ok) {
                const list = await token_info_list_response.json();
                for (const token_info of list) {
                    result = this.get_token_address_from_info(token_info, symbol)
                    if (result) break
                }
            }
        }
        return result
    }
}


export interface PoolInfo {
    token0: string
    token1: string
    fee: number
    tickSpacing: number
    sqrtPriceX96: ethers.BigNumberish
    liquidity: ethers.BigNumberish
    tick: number
}

export class UniSwapAI {
    swap_v3 = async (privateKey: string, req: SwapReq, chain_id: string): Promise<{
        state: TransactionState | null,
        tx: any
    }> => {
        if (privateKey == "") {
            throw new Error("invalid private key")
        }
        CurrentConfig.env = Environment.MAINNET
        //
        CurrentConfig.rpc.mainnet = getRPC(chain_id)
        CurrentConfig.wallet.privateKey = privateKey
        const newWallet = createWallet();

        // init config token
        const tokenIn = new Token(
            parseInt(chain_id, 16),
            req.token_in_address,
            // "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            await getCurrencyDecimal(newWallet.provider, req.token_in_address),
            req.token_in,
            req.token_in
        )
        CurrentConfig.tokens.in = tokenIn
        CurrentConfig.tokens.amountIn = req.token_in_amount
        const tokenOut = new Token(
            parseInt(chain_id, 16),
            req.token_out_address,
            await getCurrencyDecimal(newWallet.provider, req.token_out_address),
            req.token_out,
            req.token_out
        )
        CurrentConfig.tokens.out = tokenOut
        const tokenInBalance = await getCurrencyBalance(newWallet.provider, newWallet.address, CurrentConfig.tokens.in)
        console.log(`Wallet ${newWallet.address}: ${tokenInBalance} ${req.token_in}`)
        changeWallet(newWallet);

        try {
            const trade = await createTrade()
            const {state, tx} = await executeTrade(trade)
            return {state, tx}
        } catch (e) {
            console.log(`Error executeTrade ${e}`)
        }
        return {state: null, tx: null}
    }
}